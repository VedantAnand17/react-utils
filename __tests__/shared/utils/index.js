import { time } from '.';

/* Mocking up time.timer(..), because Jest mocks for timers have some issues
 * with timeouts inside promises inside async code, and such complex stuff. */
const timerSpy = jest.spyOn(time, 'timer')
  .mockImplementation(() => Promise.resolve());

jest.useFakeTimers();

const { withRetries } = require('.');

const FAILURE = 'FAILURE';
const SUCCESS = 'SUCCESS';

describe('withRetries(..)', () => {
  function newAsyncTestAction(numFailures = 0) {
    let counter = 0;
    return jest.fn(async () => {
      await time.timer(100);
      if (counter++ < numFailures) throw Error(FAILURE);
      return SUCCESS;
    });
  }

  function newSyncTestAction(numFailures) {
    let counter = 0;
    return jest.fn(() => {
      if (counter++ < numFailures) throw Error(FAILURE);
      return SUCCESS;
    });
  }

  /* eslint-disable no-await-in-loop */
  it('works correctly with async actions', async () => {
    for (let i = 0; i < 4; ++i) {
      timerSpy.mockClear();
      const action = newAsyncTestAction(i);
      await expect(withRetries(action)).resolves.toBe(SUCCESS);
      expect(action).toHaveBeenCalledTimes(i + 1);
      expect(timerSpy.mock.calls).toMatchSnapshot();
    }
    timerSpy.mockClear();
    const action = newAsyncTestAction(5);
    await expect(withRetries(action)).rejects.toThrowErrorMatchingSnapshot();
    expect(action).toHaveBeenCalledTimes(5);
    expect(timerSpy.mock.calls).toMatchSnapshot();
  });

  it('works correctly with sync actions', async () => {
    for (let i = 0; i < 4; ++i) {
      timerSpy.mockClear();
      const action = newSyncTestAction(i);
      await expect(withRetries(action)).resolves.toBe(SUCCESS);
      expect(action).toHaveBeenCalledTimes(i + 1);
      expect(timerSpy.mock.calls).toMatchSnapshot();
    }
    timerSpy.mockClear();
    const action = newSyncTestAction(5);
    await expect(withRetries(action)).rejects.toThrowErrorMatchingSnapshot();
    expect(action).toHaveBeenCalledTimes(5);
    expect(timerSpy.mock.calls).toMatchSnapshot();
  });

  it('respects custom "maxRetries" and "interval" arguments', async () => {
    for (let i = 0; i < 2; ++i) {
      timerSpy.mockClear();
      const action = newAsyncTestAction(i);
      await expect(withRetries(action, 2, 500)).resolves.toBe(SUCCESS);
      expect(action).toHaveBeenCalledTimes(i + 1);
      expect(timerSpy.mock.calls).toMatchSnapshot();
    }
    timerSpy.mockClear();
    const action = newAsyncTestAction(3);
    await expect(withRetries(action, 2, 500))
      .rejects.toThrowErrorMatchingSnapshot();
    expect(action).toHaveBeenCalledTimes(2);
    expect(timerSpy.mock.calls).toMatchSnapshot();
  });
  /* eslint-enable no-await-in-loop */
});
