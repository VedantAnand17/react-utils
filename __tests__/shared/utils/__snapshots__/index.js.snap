// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withRetries(..) respects custom "maxRetries" and "interval" arguments 1`] = `
Array [
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) respects custom "maxRetries" and "interval" arguments 2`] = `
Array [
  Array [
    100,
  ],
  Array [
    500,
  ],
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) respects custom "maxRetries" and "interval" arguments 3`] = `"FAILURE"`;

exports[`withRetries(..) respects custom "maxRetries" and "interval" arguments 4`] = `
Array [
  Array [
    100,
  ],
  Array [
    500,
  ],
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) works correctly with async actions 1`] = `
Array [
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) works correctly with async actions 2`] = `
Array [
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) works correctly with async actions 3`] = `
Array [
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) works correctly with async actions 4`] = `
Array [
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) works correctly with async actions 5`] = `"FAILURE"`;

exports[`withRetries(..) works correctly with async actions 6`] = `
Array [
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
  Array [
    1000,
  ],
  Array [
    100,
  ],
]
`;

exports[`withRetries(..) works correctly with sync actions 1`] = `Array []`;

exports[`withRetries(..) works correctly with sync actions 2`] = `
Array [
  Array [
    1000,
  ],
]
`;

exports[`withRetries(..) works correctly with sync actions 3`] = `
Array [
  Array [
    1000,
  ],
  Array [
    1000,
  ],
]
`;

exports[`withRetries(..) works correctly with sync actions 4`] = `
Array [
  Array [
    1000,
  ],
  Array [
    1000,
  ],
  Array [
    1000,
  ],
]
`;

exports[`withRetries(..) works correctly with sync actions 5`] = `"FAILURE"`;

exports[`withRetries(..) works correctly with sync actions 6`] = `
Array [
  Array [
    1000,
  ],
  Array [
    1000,
  ],
  Array [
    1000,
  ],
  Array [
    1000,
  ],
]
`;
